### ISSUE TITLE
Only Build gRPC Protobuf with the gRPC Gateway

### ISSUE THREAD

**Issue by amit-traiana** [rules-proto-grpc/rules_proto_grpc#31]
The `protoc-gen-grpc_gateway` plugin generates gRPC gateway from *gRPC* protobuf files. 
Consider this example:

**service_a.proto**
```
syntax = "proto3";

package service_a;

message TestMeServiceA {
    string name = 1;
}
```

**service_b.proto**
```
syntax = "proto3";

package service_a;

import "google/api/annotations.proto";
import "service_a.proto";

service TestService {
    rpc Echo(service_a.TestMeServiceA) returns (service_a.TestMeServiceB) {
      option (google.api.http) = {
        post: "/v1/example/echo"
        body: "*"
      }
  }
}
```

`service_a` only hold a message. `service_b` reference to `service_a`, and also define gRPC request that use the message defined in `service_a`.

When trying to build the gateway against it, Bazel will complain he can't find a file called `service_a.gw.go`. Checking the final DepSet, there's 4 files:

- service_a.go
- service_a.gw.go
- service_b.go
- service_b.gw.go

(the reason is generated two files for each proto is because the [gateway_grpc_aspect](https://github.com/rules-proto-grpc/rules_proto_grpc/blob/master/github.com/grpc-ecosystem/grpc-gateway/gateway_grpc_compile.bzl) defines both go and grpc as plugins).

However, the `gateway` plugin, will not create a `*.gw.go` file, when the file doesn't have gRPC definition (it's doesn't print any error, just doesn't create anything).

So the real files created on the filesystem would be:

- service_a.go
- service_b.go
- service_b.gw.go

That's why Bazel complains about missing `service_.gw.go`.

**Comment by amit-traiana** [comment]
Hi again,

I came to the point where I need to use the gRPC gateway, and I decided because I'm already deeply invested in Bazel - it's time to practice some Skylark and help out if I can. I tried to debug the original issues, and found the real problem - I modified the original message and title to reflect the problem.

I'm assuming this never works, because the aspect recursively create every file into the final depset across all rules.

That being said, `proto_compile_aspect_impl` looks like the heart of the project, and I figured I should consult with you before attempting to find a solution myself. Perhaps you already aware of it, or thought of a way to handle it.

Thanks!

**Comment by aaliddell** [comment]
Thanks for the report and sorry for the delay. Now that I've got 1.0.0 out, I should have a bit more time to look at these bugs. There might be a quick solution in changing gateway to use the directory output mode, which is helpful when the outputs can't be predicted.

**Comment by amit-traiana** [comment]
No worries. Thanks for the time your investing in the project!
Happy to hear 1.0.0 is out! I'll switch to master. 

Is this a settings you can change per plug-in? because I couldn't find anything that defines the output mode when checking the plugin BUILD file. If you can point me to the specific spot, I can change it and do the testing here to make sure it indeed solves the issue.

**Comment by aaliddell** [comment]
I've tried enabling it on the 'test-gateway' branch: https://github.com/rules-proto-grpc/rules_proto_grpc/tree/test-gateway

Unfortunately, rules_go does not support being passed a directory to its `srcs`, meaning it'll presently probably fail. There appears to be ongoing discussion about adding directory support at https://github.com/bazelbuild/rules_go/pull/2271

**Comment by amit-traiana** [comment]
Oh, I see now. I was sure `output_directory` is a Bazel parameter. 

> Unfortunately, rules_go does not support being passed a directory to its srcs, meaning it'll presently probably fail. There appears to be ongoing discussion about adding directory support at bazelbuild/rules_go#2271

Something else happening for me:
```
cp: cannot stat 'bazel-out/k8-fastbuild/bin/external/com_github_scalapb_scalapb_repo/scalapb_proto/gateway_grpc_compile_aspect_verb0/_plugin_grpc_gateway_plugin/*': No such file or directory
```

It looks like a folder is being created for every target. However, some of the targets are normal protobuf targets, and not gRPC protobuf targets. It means some folders will end up being empty, because running the gRPC plug-in will output nothing. 

Then when `cp -r` part fails, because `cp` from an empty folder with '*' will not work.

Thanks!

**Comment by aaliddell** [comment]
Ah interesting, that's a separate bug. Most plugins always generate a file, regardless of whether it's empty or not. I've opened a new issue for that here: https://github.com/rules-proto-grpc/rules_proto_grpc/issues/39

**Comment by aaliddell** [comment]
#39 is resolved and I've added the fix to the test-gateway branch, so you should now see the rules_go issue.

**Comment by amit-traiana** [comment]
#41

**Comment by amit-traiana** [comment]
@aaliddell I came to a point I need to use the Gateway - so I have invested some time trying to figure what's exactly is broken here. I don't have all the answer yet, but I want to share some information with you, maybe something will click as you are much more familiar with the code. 

So, previously, we added `output_directory = True`, as the Gateway plug-in will only create output files for protobuf files with gRPC service in it (so we can't tell the final output list). With `output_directory = True`, I see the following:

**gateway_grpc_compile:**
It fails during copy here:

```
        # Copy directories and files to shared output directory in one action
        ctx.actions.run_shell(
            mnemonic = "CopyDirs",
            inputs = output_dirs + command_input_files,
            outputs = [new_dir],
            command = " && ".join(command_parts),
            progress_message = "copying directories and files to {}".format(new_dir.path),
        )
```
The reason it fails, is because some directories doesn't exists (and `cp` can't copy a file to a folder that doesn't exists). So if for example I have a transitive dependency, that transitive dependency will be built in `bazel-out/k8-fastbuild/project/transitive_dep/transitive.go`. But when it tried to copy it to the target folder (`bazel-out/k8-fastbuild/project/gateway_target/transitive_dep`) it fails because while `project/transitive_dep` exists, `project/gateway_target/transitive_depo` isn't.  So I'm not yet sure why it won't work, but I was able to hack my way around it by just making sure the folder the file is being copied too is being created if doesn't exists:

```
                    new_dir_path = "{}/{}".format(new_dir.path, path)
                    command_parts.append("mkdir -p '{}' && cp '{}' '{}'".format(
                        new_dir_path,
                        file.path,
                        new_dir_path,
                    ))
```

This works, and `gateway_grpc_compile` output working Go gateway file. There's some warning that print out that might be connected to the problem (they were also showing prior to the hacky fix), but I ignored them for now being just "warning" (they doesn't fail the build process):

`cp: warning: source directory 'bazel-out/k8-fastbuild/bin/external/scalapb_proto/protobuf/scalapb/scalapb_proto/gateway_grpc_compile_aspect_verb0/_plugin_grpc_gateway_plugin/.' specified more than once`

So while `gateway_grpc_compile` works with this magical hack - I'm really interested in `gateway_grpc_library` (which depends on `compile` of-course). 

There's something strange going on here. Calling the `gateway_grpc_library` will fail:

```
Traceback (most recent call last):
        File "/home/amitz/git/depo/cme/orgmodel/BUILD", line 92
                go_library(name = 'orgmodel_gateway_api')
        File "/home/amitz/.cache/bazel/_bazel_amitz/07f508f89225552f1f462fc4df3f5714/external/io_bazel_rules_go/go/private/rules/library.bzl", line 42, in _go_library_impl
                go.archive(go, source)
        File "/home/amitz/.cache/bazel/_bazel_amitz/07f508f89225552f1f462fc4df3f5714/external/io_bazel_rules_go/go/private/actions/archive.bzl", line 48, in go.archive
                split_srcs(source.srcs)
        File "/home/amitz/.cache/bazel/_bazel_amitz/07f508f89225552f1f462fc4df3f5714/external/io_bazel_rules_go/go/private/common.bzl", line 110, in split_srcs
                fail(<1 more arguments>)
Unknown source type api
```

The `Unknown source type api` it reference, is actually the importpath parameter `github.com/bla/api`. For some reason the Go library plugin get it as input, and complains is not a go file. However, if I call the Macro you create directly (so if the target name is "my_test" I `bazel build //:my_test_pb`), everything passes. I can see the output `gw.go` files, and they are correct and working properly. I'm not sure yet why.

Anything makes sense to you? I can supply a demo repo if you wish to see the failures yourself. 
Thanks!

**Comment by aaliddell** [comment]
I'll have to go back and look at the folder issue, as I would expect we'd see that problem already with other languages that are already set to use `output_directory = True`. Unless we're missing an edge case in the set of workspace tests, which is highly likely.

As for the two questions:

1. Yep, that's the rules_go issue that's now finally surfaced. The code in question is [here](https://github.com/bazelbuild/rules_go/blob/2a0e3a07e9ed9aa9b7afd1a222638ba52166e52d/go/private/common.bzl#L82-L112). As you can see, it's trying to split the inputs by the extension, but the directory input `/some/path/api` does not have an extension and it prints the basename, which is `api` in this case.

2. I'm not sure it did. It uses a [list of expected output suffixes](https://github.com/stackb/rules_proto/blob/1d6b84118399828511faeecc145d399c1e7bdee2/github.com/grpc-ecosystem/grpc-gateway/BUILD.bazel#L5-L7), so would fail already when the proto plugin fails to create outputs for empty files, which we use `output_directory` to work around.

Solving this is possibly very difficult, as you either need to either:

A. Get the plugin to make all outputs, even if empty
B. Get rules_go to accept directory inputs

A third option may be to wrap the proto plugin and force generate the empty files, allowing you to disable `output_directory`.

**Comment by aaliddell** [comment]
The reason the files are not generated by the plugin for service-less files appears to be here: 

https://github.com/grpc-ecosystem/grpc-gateway/blob/d1093dfe831a1910372435673ff1b4aa3bf776b2/protoc-gen-grpc-gateway/internal/gengateway/generator.go#L96-L99

**Comment by amit-traiana** [comment]
Sure, let me know if you figured what causing the non-created folder issue.

> A third option may be to wrap the proto plugin and force generate the empty files, allowing you to disable output_directory.
That's what I'm currently looking at. If the Protobuf has no gRPC in it, the Buffer printed to stdout will be empty, causing `protoc` to not create a file (so it's actually `protoc` who don't create empty files`).  What I'm currently trying is to always print a comment. So if it's empty, it's an empty file with just comment, and if not - it's a comment + the Gateway code. Let's see if I can make it work. It does means I have to work with a Fork or ask the Gateway guys to added this (which I'm assuming they won't like).

**Comment by amit-traiana** [comment]
Hey again.

**Comment by aaliddell** [comment]
Hmm, Ok. I'll tag this issue as blocked by those upstream issues, as I don't see a way forward without at least one of those being resolved.

**Comment by aaliddell** [comment]
This should be resolved in 3.0.0, where non-transitive compilation is available and missing generated files are 'filled'
